### PR TITLE
Fix Slice Folder OOB Crash and onnx.Shape lowering

### PR DESCRIPTION
**Pull request by zjgarvey**
1. Clamps OOB start index to 0 in slice folder
2. Adds a more descriptive `emitError` in slice folder if the creation of the `DenseElementsAttr` would fail due to a bad result shape.
3. Fixes the `onnx.Shape` lowering to default to `inputRank` for `end` instead of `-1`. When `end==-1` the last element was missing when slicing.

### CODE REVIEW

File: lib/Dialect/Torch/IR/TorchOps.cpp [hunk 1]
```diff
@@ -3998,6 +3998,7 @@ OpFoldResult AtenSliceTensorOp::fold(FoldAdaptor adaptor) {
     int64_t limit = end.getValue().getSExtValue();
     int64_t stride = step.getValue().getSExtValue();
     begin = begin < 0 ? begin + inType.getSizes()[dimInt] : begin;
+    begin = std::max(begin, (int64_t)0);
```
Comment (rsuderman) [inline lib/Dialect/Torch/IR/TorchOps.cpp:4001]: Nit: you can do `std::max<int64_t>(begin, 0)` I think

File: lib/Dialect/Torch/IR/TorchOps.cpp [hunk 2]
```diff
@@ -4038,6 +4039,14 @@ OpFoldResult AtenSliceTensorOp::fold(FoldAdaptor adaptor) {
       }
     };
     recursiveIter(recursiveIter, 0, 0);
+    if ((int64_t)values.size() != count) {
```
Comment (rsuderman) [inline lib/Dialect/Torch/IR/TorchOps.cpp:4042]: Try to use `static_cast` so it should be

```
    if (static_cast<int64_t>(values.size()) != count) {
```

**Review by rsuderman — APPROVED** [review]

**Review by vinayakdsci — APPROVED** [review]
LGTM!

File: lib/Dialect/Torch/IR/TorchOps.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (vinayakdsci) [reply lib/Dialect/Torch/IR/TorchOps.cpp]: This is resolved, thanks!

**Pull request merged**